File: lib/JSONRPC/filters/client/SignatureAdd.php
<?php
require_once("JSONRPC/ClientFilterBase.php");

/**
* JSON-RPC 2.0 client filter plugin.
* Adds authentication and signed request expiration for the JSONRPC ResellerDomainsAPI.
* Also translates thrown exceptions.
*/
class JSONRPC_filter_signature_add extends JSONRPC_client_filter_plugin_base
{
	protected $_strAPIKEY;
	protected $_arrExtraURLVariables;
	
	
	public function __construct($strAPIKEY, $arrExtraURLVariables=array())
	{
		$this->_strAPIKEY=$strAPIKEY;
		$this->_arrExtraURLVariables=$arrExtraURLVariables;
	}
	
	
	public function beforeJSONEncode(&$arrRequest)
	{
		$arrRequest["expires"]=time()+86400;
	}
	
	
	public function afterJSONEncode(&$strJSONRequest, &$strEndpointURL, &$arrHTTPHeaders)
	{
		static $strHashingFunction=NULL;
		
		if(!isset($strHashingFunction))
		{
			if(in_array("sha256", hash_algos()))
				$strHashingFunction="sha256";
			else if(in_array("md5", hash_algos()))
				$strHashingFunction="md5";
			else
				throw new Exception("hash_algos() does not list at least one of the required hashing functions: md5 or sha256.");
		}
		
		$strVerifyHash=hash_hmac($strHashingFunction, $strJSONRequest, $this->_strAPIKEY);
		
		if(strpos($strEndpointURL, "?")!==false)
			$strEndpointURL.="&";
		else
			$strEndpointURL.="?";
		
		//Does not check for existing URL params. Either way, duplication or overwriting must be considered a conflict and must be resolved.
		$strEndpointURL.="verify=".urlencode($strVerifyHash);
		
		foreach($this->_arrExtraURLVariables as $strKey=>$strValue)
			$strEndpointURL.="&".urlencode($strKey)."=".urlencode($strValue);
	}
}

File: lib/JSONRPC/filters/client/SignatureAdd.py
"""
* JSON-RPC 2.0 client filter plugin.
* Adds authentication and signed request expiration for the JSONRPC ResellerDomainsAPI.
* Also translates thrown exceptions.
"""

import hmac
import json
import urllib2
import sys
sys.path.append("../..")
import urllib
import hashlib
import time
from ...ClientFilterBase import JSONRPC_client_filter_plugin_base
from pprint import pprint

class JSONRPC_filter_signature_add(JSONRPC_client_filter_plugin_base):


	"""
	* Private key used for hashed messages sent to the server
	"""
	strAPIKey = ""


	"""
	* Extra URL variables
	"""
	dictExtraURLVariables = {}


	"""
	* Private key used for hashed messages sent to the server
	"""
	strKeyMetaData = ""


	"""
	* This is the constructor function. It creates a new instance of JSONRPC_filter_signature_add.
	* Example: JSONRPC_filter_signature_add("secretKey")
	*
	* @param string strKEY. The private key used for hashed messages sent to the server.
	"""

	def __init__(self, strKey, dictExtraURLVariables):

		self.strAPIKey = strKey
		self.dictExtraURLVariables = dictExtraURLVariables
		self.getKeyMetaData()

	def getKeyMetaData (self) :
		strKEYSplit = self.strAPIKey.split(":", 2)
		self.strKeyMetaData = strKEYSplit[0]


	"""
	* This function sets an uptime for the request.
	*
	* @param dictionary dictFilterParams. It is used for reference return for multiple variables,
	* which can be retrieved using specific keys
	* -"dictRequest"
	*
	* @return dict dictFilterParams
	"""
	def beforeJSONEncode(self, dictFilterParams):

		dictFilterParams["dictRequest"]["expires"] = int(time.time()+86400)

		return dictFilterParams



	"""
	* This function is used for authentication. It alters the Endpoint URL such that it contains
	* a specific signature.
	* 
	* @param dictionary dictFilterParams. It is used for reference return for multiple variables,
	* which can be retrieved using specific keys
	* - "strJSONRequest"
	* - "strEndpointURL"
	* - "dictHTTPHeaders"
	*
	* @return dict dictFilterParams
	"""
	def afterJSONEncode(self, dictFilterParams):

		strVerifyHash = hmac.new(self.strAPIKey, dictFilterParams["strJSONRequest"], hashlib.md5).hexdigest()

		
		if self.strKeyMetaData.isdigit():
		    strVerifyHash = self.strKeyMetaData + ":" + strVerifyHash
		else:
			raise Exception("Invalid API key format")

		
		if dictFilterParams["strEndpointURL"].find('?') != -1:
			dictFilterParams["strEndpointURL"] += "&"

		else:
			dictFilterParams["strEndpointURL"] += "?"

		if dictFilterParams["strEndpointURL"].find("?verify") == -1:
			dictFilterParams["strEndpointURL"] += "verify="+urllib.quote(strVerifyHash);

		for key, value in self.dictExtraURLVariables.items():
			value = str(value)
			dictFilterParams["strEndpointURL"] += "&"+urllib.quote(key)+"="+urllib.quote(value)		

		
		return dictFilterParams

		
